internal/enum/enum.go

```diff
@@ -0,0 +1,353 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// enum is a helper for generating boilerplate related to Go enums.
+//
+// To generate boilerplate for a given file, use
+//
+//	//go:generate go run github.com/bufbuild/protocompile/internal/enum
+//
+// To generate code for an enum type, annotate it with the appropriate
+// directives.
+//
+//   - //enum:string Name will generate a stringifying methods named Name,
+//     which default to "String".
+//
+//   - //enum:gostring Name is similar, but the default name is "GoString"
+//     and each constant stringifies to "package.Constant".
+//
+//   - //enum:fromstring Name generates a function that takes a string and
+//     returns the constant that stringifies to that value, or 0 if there is
+//     no such value.
+//
+//   - //enum:doc directive "..." adds a line of doc comment to the method
+//     generated by directive with the given text.
+//
+//   - //enum:import "..." will include a package to import in the generated
+//     file.
+//
+//   - //enum:skip, attached to an enum constant, will cause that constant to
+//     be ignored by the generator.
+//
+//   - //enum:stringfunc Name provides a function to use for stringifying
+//     based on the name of a constant (see below).
+//
+// The string representation of an enum constant is computed by the following
+// process:
+//
+//  1. If the constant is annotated with //enum:string "...", it will use that
+//     value.
+//
+//  2. If the enum type is annotated with //enum:stringfunc Name, the string
+//     will be Name("Constant"), where Constant is the Go-declared name of the
+//     constant.
+//
+//  3. The string is "Constant".
+package main
+
+import (
+	"debug/buildinfo"
+	_ "embed"
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"os"
+	"slices"
+	"strconv"
+	"strings"
+	"text/scanner"
+	"text/template"
+	"unicode"
+
+	"github.com/bufbuild/protocompile/internal/ext/slicesx"
+)
+
+var (
+	//go:embed generated.go.tmpl
+	tmplText string
+	tmpl     = template.Must(template.New("generated.go.tmpl").Parse(tmplText))
+)
+
+type Directive struct {
+	Pos  token.Pos
+	Name string
+	Args []string
+}
+
+func HasDirectives(comments []*ast.CommentGroup) bool {
+	for _, g := range comments {
+		for _, c := range g.List {
+			rest, ok := strings.CutPrefix(c.Text, "//enum:")
+			if ok && rest != "" {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func ParseDirectives(fs *token.FileSet, comments []*ast.CommentGroup) ([]Directive, error) {
```

Comment:
These seems like _serious_ overkill. How about just define the enums (name, doc, constants) in a yaml file and then parsing the YAML and generating Go code (including for the type definition and constants) is trivial. I don't think it needs to be very configurable -- you could make it always emit some sort of "Lookup<Type>" factory function or maybe just make it a simple bool flag in the yaml.

Reply:
Lol using a config file hadn't occurred to me.